xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,785 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+		// If the address is ejected, and the current time is after
+		// ejection_timestamp + min(base_ejection_time (type: time.Time) *
+		// multiplier (type: int), max(base_ejection_time (type: time.Time),
+		// max_ejection_time (type: time.Time))), un-eject the address.
+		if !addrInfo.latestEjectionTimestamp.IsZero() && curTimeAfterEt {
+			b.unejectAddress(addrInfo)
+		}
+	}
+
+	// This conditional only for testing (since the interval timer algorithm is
+	// called manually), will never hit in production.
+	if b.intervalTimer != nil {
+		b.intervalTimer.Stop()
+	}
+	b.intervalTimer = afterFunc(b.cfg.Interval, func() {
+		b.intervalTimerAlgorithm()
+	})
+}
+
+// numAddrsWithAtLeastRequestVolume returns the number of addresses present in
+// the map that have request volume of at least requestVolume.
+func (b *outlierDetectionBalancer) numAddrsWithAtLeastRequestVolume() uint32 {
+	var numAddrs uint32
+	for _, addrInfo := range b.addrs {
+		if uint32(addrInfo.callCounter.inactiveBucket.requestVolume) >= b.cfg.SuccessRateEjection.RequestVolume {
+			numAddrs++
+		}
+	}
+	return numAddrs
+}
+
+// meanAndStdDevOfSuccessesAtLeastRequestVolume returns the mean and std dev of
+// the number of requests of addresses that have at least requestVolume.
+func (b *outlierDetectionBalancer) meanAndStdDevOfSuccessesAtLeastRequestVolume() (float64, float64) {
+	// 2. Calculate the mean and standard deviation of the fractions of
+	// successful requests among addresses with total request volume of at least
+	// success_rate_ejection.request_volume.
+	var totalFractionOfSuccessfulRequests float64
+	var mean float64
+	for _, addrInfo := range b.addrs {
+		// "of at least success_rate_ejection.request_volume"
+		if uint32(addrInfo.callCounter.inactiveBucket.requestVolume) >= b.cfg.SuccessRateEjection.RequestVolume {
+			totalFractionOfSuccessfulRequests += float64(addrInfo.callCounter.inactiveBucket.numSuccesses) / float64(addrInfo.callCounter.inactiveBucket.requestVolume)
+		}
+	}
+	mean = totalFractionOfSuccessfulRequests / float64(len(b.addrs))
+	var sumOfSquares float64
+	for _, addrInfo := range b.addrs {
+		devFromMean := (float64(addrInfo.callCounter.inactiveBucket.numSuccesses) / float64(addrInfo.callCounter.inactiveBucket.requestVolume)) - mean
+		sumOfSquares += devFromMean * devFromMean
+	}
+
+	variance := sumOfSquares / float64(len(b.addrs))
+	return mean, math.Sqrt(variance)
+
+}
+
+// syccessRateAlgorithm ejects any addresses where the success rate falls below
+// the other addresses according to mean and standard deviation, and if overall
+// applicable from other set heuristics.
+func (b *outlierDetectionBalancer) successRateAlgorithm() {
+	// 1. If the number of addresses with request volume of at least
+	// success_rate_ejection.request_volume is less than
+	// success_rate_ejection.minimum_hosts, stop.
+	if b.numAddrsWithAtLeastRequestVolume() < b.cfg.SuccessRateEjection.MinimumHosts { // TODO: O(n) search, is there a way to optimize this?
```

Comment:
I think it's fine (remove TODO?) since the number of backends should generally be very small (<10), right?

(Otherwise, we could optimize by keeping a running count that's updated dynamically.)

Reply:
Yeah good suggestion, I had something like that with numAddrsEjected. I don't know how many backends in practice, but this seems fine, and adding a dynamic count doesn't seem worth it. Thanks.